sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -716,6 +716,17 @@ object SimplifyCasts extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Simplify if clauses of pattern `if(p, null, true|false)`, by replacing them with
+ * AND or OR clauses which are simpler and can better be pushed down
+ */
+object SimplifyIf extends Rule[LogicalPlan] {
+  val nullLiteral = Literal(null, BooleanType)
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+    case If(p, Literal(null, _), FalseLiteral) => And(p, nullLiteral)
+    case If(p, Literal(null, _), TrueLiteral) => Or(p, nullLiteral)
```

Comment:
Is it common enough to write an if with a `null` with returning a bool? Adding a rule isn't free and it gives some overhead to all the queries basically during the planning. I would avoid doing it if it's pretty unlikely.

Reply:
How about merging these to the existing `SimplifyConditionals` as mentioned by @tanelk above? it will then just be a few more pattern matchings, right?